pkg/client/apiutil/dynamicrestmapper.go

```diff
@@ -0,0 +1,216 @@
+	if err != nil {
+		return err
+	}
+	drm.delegate = restmapper.NewDiscoveryRESTMapper(gr)
+	return nil
+}
+
+func noKindMatchError(err error) bool {
+	_, ok := err.(*meta.NoKindMatchError)
+	return ok
+}
+
+// reload reloads the delegated RESTMapper, and will return an error only
+// if a rate limit has been hit.
+func (drm *DynamicRESTMapper) reload() error {
+	if err := drm.limiter.checkRate(); err != nil {
+		return err
+	}
+	if err := drm.setDelegate(); err != nil {
+		utilruntime.HandleError(err)
+	}
+	return nil
+}
+
+func (drm *DynamicRESTMapper) KindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
+	gvk, err := drm.delegate.KindFor(resource)
+	if noKindMatchError(err) {
+		if rerr := drm.reload(); rerr != nil {
+			return schema.GroupVersionKind{}, rerr
+		}
+		gvk, err = drm.delegate.KindFor(resource)
+	}
+	return gvk, err
+}
+
+func (drm *DynamicRESTMapper) KindsFor(resource schema.GroupVersionResource) ([]schema.GroupVersionKind, error) {
+	gvks, err := drm.delegate.KindsFor(resource)
+	if noKindMatchError(err) {
+		if rerr := drm.reload(); rerr != nil {
+			return nil, rerr
+		}
+		gvks, err = drm.delegate.KindsFor(resource)
+	}
+	return gvks, err
+}
+
+func (drm *DynamicRESTMapper) ResourceFor(input schema.GroupVersionResource) (schema.GroupVersionResource, error) {
+	gvr, err := drm.delegate.ResourceFor(input)
+	if noKindMatchError(err) {
+		if rerr := drm.reload(); rerr != nil {
+			return schema.GroupVersionResource{}, rerr
+		}
+		gvr, err = drm.delegate.ResourceFor(input)
+	}
+	return gvr, err
+}
+
+func (drm *DynamicRESTMapper) ResourcesFor(input schema.GroupVersionResource) ([]schema.GroupVersionResource, error) {
+	gvrs, err := drm.delegate.ResourcesFor(input)
+	if noKindMatchError(err) {
+		if rerr := drm.reload(); rerr != nil {
+			return nil, rerr
+		}
+		gvrs, err = drm.delegate.ResourcesFor(input)
+	}
+	return gvrs, err
+}
+
+func (drm *DynamicRESTMapper) RESTMapping(gk schema.GroupKind, versions ...string) (*meta.RESTMapping, error) {
+	m, err := drm.delegate.RESTMapping(gk, versions...)
+	if noKindMatchError(err) {
+		if rerr := drm.reload(); rerr != nil {
+			return nil, rerr
+		}
+		m, err = drm.delegate.RESTMapping(gk, versions...)
+	}
+	return m, err
+}
+
+func (drm *DynamicRESTMapper) RESTMappings(gk schema.GroupKind, versions ...string) ([]*meta.RESTMapping, error) {
+	ms, err := drm.delegate.RESTMappings(gk, versions...)
+	if noKindMatchError(err) {
+		if rerr := drm.reload(); rerr != nil {
+			return nil, rerr
+		}
+		ms, err = drm.delegate.RESTMappings(gk, versions...)
+	}
+	return ms, err
+}
+
+func (drm *DynamicRESTMapper) ResourceSingularizer(resource string) (singular string, err error) {
+	s, err := drm.delegate.ResourceSingularizer(resource)
+	if noKindMatchError(err) {
+		if rerr := drm.reload(); rerr != nil {
+			return "", rerr
+		}
+		s, err = drm.delegate.ResourceSingularizer(resource)
+	}
+	return s, err
+}
+
+type limiter struct {
```

Comment:
nit: less generic name for this?  maybe `discoveryLimitter`?

Reply:
it's actually fairly generic -- just something to make it obvious why we need something besides `rate.Limitter`